src/cirrus/lib2/events.py

```diff
@@ -0,0 +1,149 @@
+import json
+import os
+from contextlib import AbstractContextManager, contextmanager
+from logging import Logger, getLogger
+from typing import Dict
+
+import boto3
+
+from .enums import StateEnum
+from .statedb import StateDB
+from .utils import SNSPublisher
+
+logger = getLogger(__name__)
+
+
+class WorkflowEventManager:
+    """A class for managing payload state change events, including storage of
+    state(dynamo), activity(timestream), and notifications of decisions and/or status
+    changes (SNS)."""
+
+    def __init__(
+        self,
+        logger: Logger = logger,
+        boto3_session: boto3.Session = None,
+        statedb: StateDB = None,
+        batch_size: int = 10,
+    ):
+        self.logger = logger
+        self._boto3_session = boto3_session if boto3_session else boto3.Session()
+        wf_event_topic_arn = os.getenv("CIRRUS_WORKFLOW_EVENT_TOPIC_ARN", None)
+        self.event_publisher = (
+            SNSPublisher(
+                wf_event_topic_arn,
+                logger=logger,
+                boto3_session=self._boto3_session,
+                batch_size=batch_size,
+            )
+            if wf_event_topic_arn
+            else None
+        )
+        self.statedb = (
+            statedb if statedb is not None else StateDB(session=self._boto3_session)
+        )
+
+    def flush(self):
+        if self.event_publisher:
+            self.event_publisher.execute()
+
+    @classmethod
+    @contextmanager
+    def handler(
+        cls: "WorkflowEventManager", *args, **kwargs
+    ) -> AbstractContextManager["WorkflowEventManager"]:
+        wfem = cls(*args, **kwargs)
+        try:
+            yield wfem
+        finally:
+            wfem.flush()
+
+    def announce(
+        self,
+        event_type: str,
+        payload: dict = None,
+        payload_id: str = None,
+        extra_message: Dict = None,
+    ) -> None:
+        """Construct message payload and publish to WorkflowEventTopic"""
+        if self.event_publisher is None:
+            return
+
+        if payload is None:
+            if payload_id is None:
+                raise ValueError("must specify payload_id or payload")
+        elif payload_id is None:
+            payload_id = payload["id"]
+
+        if payload is not None and payload.get("id", payload_id) != payload_id:
+            raise ValueError(
+                "payload_id and payload['id'] must match, if both supplied."
+            )
+        message = {
+            "event_type": event_type,
+            "payload": payload,
+            "payload_id": payload_id,
+        }
+
+        if extra_message:
+            if not all(k not in message for k in extra_message):
+                raise ValueError(
+                    "extra_message parameters must not include "
+                    "event_type, payload, or payload_id."
+                )
```

Comment:
The `payload_id` here is renamed to `id` in the subsequent PR #263, and the error message is made dynamic based on the keys of the `message`.